File: VB/Customfigs_VB/VectorDrawCustomFigs_VB/VectorDrawArrowLine.vb
Imports System
Imports System.Runtime.InteropServices
Imports System.ComponentModel
Imports VectorDraw.Serialize
Imports VectorDraw.Professional
Imports VectorDraw.Professional.vdPrimaries
Imports VectorDraw.Professional.vdFigures
Imports VectorDraw.Professional.vdCollections
Imports VectorDraw.Professional.vdObjects
Imports VectorDraw.Professional.Constants
Imports VectorDraw.GeomeTry
Imports VectorDraw.Render
Imports VectorDraw.Professional.Actions
Imports VectorDraw.Actions

'namespace 
Public Class VectorDrawArrowLine
    Inherits vdFigure ', IvdProxyFigure
    Implements IvdProxyFigure

#Region "Private fields"
    Private mArrowPoints As gPoints = New gPoints()
    Private mArrowSize As Double = 1D
    Private mStartPoint As gPoint = New gPoint()
    Private mEndPoint As gPoint = New gPoint()
#End Region

#Region "ctor"
    Public Sub New()
        mArrowPoints.Add(New gPoint(0D, 0D, 0D))
        mArrowPoints.Add(New gPoint(-1D, -0.1665D, 0D))
        mArrowPoints.Add(New gPoint(-1D, 0.1665D, 0D))
        mArrowPoints.Add(New gPoint(0D, 0D, 0D))
    End Sub
#End Region

#Region "Public properties"
    'You should use the AddHistory command in your properties so any change will be writen to the history
    'The string in the AddHistory command is the name of the property exactly and should be called before the property change.
    Public Property arrowSize() As Double
        Get
            Return mArrowSize
        End Get
        Set(ByVal Value As Double)
            AddHistory("arrowSize", Value)
            mArrowSize = Value
        End Set
    End Property
    Public Property StartPoint() As gPoint
        Get
            Return mStartPoint
        End Get
        Set(ByVal Value As gPoint)
            AddHistory("StartPoint", Value)
            mStartPoint.CopyFrom(Value)
        End Set
    End Property
    Public Property EndPoint() As gPoint
        Get
            Return mEndPoint
        End Get
        Set(ByVal Value As gPoint)
            AddHistory("EndPoint", Value)
            mEndPoint.CopyFrom(Value)
        End Set
    End Property
#End Region

#Region "private methods"


    Private Function ArrowEcsMatrix(ByVal ViewDir As Vector) As Matrix
        Dim mArrowMatrix As Matrix = New Matrix()
        mArrowMatrix.ScaleMatrix(mArrowSize, mArrowSize, 1D)
        mArrowMatrix.ApplyECS2WCS(ViewDir, New Vector(StartPoint, EndPoint))
        mArrowMatrix.TranslateMatrix(EndPoint)
        Return mArrowMatrix

    End Function
#End Region
#Region "pline overrides"
    'Note that not all values/objects can be saved/serialize by VDF, and you need to write code for such properties in serialize/deserialize override of this object. 
    'You can set value of custom property to any of the following type values: 
    '     String, Double, Int32, Boolean, Byte, Int16, Int64, UInt16, UInt64, UInt32, Single, ushort, System.Color, VectorDraw.Serialize.IVDHandle and IVDSerialise
    Public Overrides Sub Serialize(ByVal serializer As Serializer)
        MyBase.Serialize(serializer)
        serializer.Serialize(mArrowSize, "arrowSize")
        serializer.Serialize(mStartPoint, "StartPoint")
        serializer.Serialize(mEndPoint, "EndPoint")
    End Sub
    Public Overrides Function DeSerialize(ByVal deserializer As DeSerializer, ByVal fieldname As String, ByVal value As Object) As Boolean
        If MyBase.DeSerialize(deserializer, fieldname, value) Then
            Return True
        ElseIf (fieldname = "arrowSize") Then
            mArrowSize = CType(value, Double)
        ElseIf (fieldname = "StartPoint") Then
            mStartPoint = CType(value, gPoint)
        ElseIf (fieldname = "EndPoint") Then
            mEndPoint = CType(value, gPoint)
        Else
            Return False
        End If
        Return True
    End Function
    Public Overrides ReadOnly Property BoundingBox() As Box
        Get
            If Not mBoundBox.IsEmpty Then
                Return mBoundBox
            End If
            mBoundBox = New Box()
            mBoundBox.AddPoint(StartPoint)
            mBoundBox.AddPoint(EndPoint)
            Dim pts As gPoints = New gPoints()
            pts.AddRange(mArrowPoints)
            ArrowEcsMatrix(New Vector(0, 0, 1)).Transform(pts)
            mBoundBox.AddPoints(pts)
            Return mBoundBox

        End Get
    End Property
    Public Overrides Sub Transformby(ByVal mat As Matrix)
        StartPoint = mat.Transform(StartPoint)
        EndPoint = mat.Transform(EndPoint)
        MyBase.Transformby(mat)
    End Sub
    Public Overrides Function Draw(ByVal render As vdRender) As vdRender.DrawStatus
        Dim doDraw As vdRender.DrawStatus = MyBase.Draw(render)
        If doDraw = vdRender.DrawStatus.Successed Then

            render.DrawLine(Me, StartPoint, EndPoint)

            render.PushMatrix(ArrowEcsMatrix(render.ViewDir))
            render.DrawSolidPolygon(Me, mArrowPoints, vdRender.PolygonType.Simple)
            render.PopMatrix()


        End If
        AfterDraw(render)
        Return doDraw
    End Function
    Public Overrides Function GetGripPoints() As gPoints
        Dim ret As gPoints = New gPoints()
        ret.Add(StartPoint)
        ret.Add(EndPoint)
        Return ret
    End Function
    Public Overrides Sub MoveGripPointsAt(ByVal Indexes As Int32Array, ByVal dx As Double, ByVal dy As Double, ByVal dz As Double)
        If Indexes Is Nothing Or Indexes.Count = 0 Or (dx = 0.0 And dy = 0.0 And dz = 0.0) Then
            Return
        End If
        Dim grips As gPoints = GetGripPoints()
        If Indexes.Count = grips.Count Then
            Dim mat As Matrix = New Matrix()
            mat.TranslateMatrix(dx, dy, dz)
            Transformby(mat)
        Else
            Dim index As Integer
            For Each index In Indexes
                Select Case index
                    Case 0
                        StartPoint += New gPoint(dx, dy, dz)
                        Exit For
                    Case 1
                        EndPoint += New gPoint(dx, dy, dz)
                        Exit For
                    Case Else
                        Exit For
                End Select
            Next
        End If
        Update()

    End Sub
    Public Overrides Sub MatchProperties(ByVal _from As VectorDraw.Professional.vdObjects.vdPrimary, ByVal thisdocument As VectorDraw.Professional.vdObjects.vdDocument)
        MyBase.MatchProperties(_from, thisdocument)
        Dim from1 As VectorDrawArrowLine = _from
        If from1 Is Nothing Then
            Return
        End If
        StartPoint = from1.StartPoint
        EndPoint = from1.EndPoint
        arrowSize = from1.arrowSize
    End Sub
    Public Overrides Function Explode() As vdEntities
        Dim Entities As vdEntities = New vdEntities()
        Entities.SetUnRegisterDocument(Document)
        If Not Document Is Nothing Then
            Document.UndoHistory.PushEnable(False)
        End If
        Dim line As vdLine = New vdLine()
        line.StartPoint = StartPoint
        line.EndPoint = EndPoint
        line.MatchProperties(Me, Document)
        Entities.AddItem(line)

        Dim pl As vdPolyline = New vdPolyline()
        pl.VertexList.AddRange(Me.mArrowPoints)
        pl.Flag = VdConstPlineFlag.PlFlagCLOSE
        pl.HatchProperties = New vdHatchProperties()
        pl.HatchProperties.FillMode = VdConstFill.VdFillModeSolid
        pl.HatchProperties.FillColor.ByBlock = True
        pl.Transformby(ArrowEcsMatrix(New Vector(0, 0, 1)))
        pl.MatchProperties(Me, Document)
        Entities.AddItem(pl)
        If Not Document Is Nothing Then
            Document.UndoHistory.PopEnable()
        End If
        Return Entities
    End Function
    Public Overrides Function IsTableObjectDependOn(ByVal table As VectorDraw.Professional.vdObjects.vdPrimary) As Boolean
        ''we call only the base check because there are not tables objects reference to this object properties.
        If Deleted = True Then Return False
        If MyBase.IsTableObjectDependOn(table) = True Then Return True
        Return False
    End Function
#End Region
End Class



Public Class ActionArrowLine
    Inherits ActionEntity
    Private line As VectorDrawArrowLine

    Private Sub New(ByVal reference As gPoint, ByVal layout As vdLayout)
        MyBase.New(reference, layout)
        line = New VectorDrawArrowLine

        line.SetUnRegisterDocument(layout.Document)
        line.setDocumentDefaults()
        line.StartPoint = reference
        line.EndPoint = reference
    End Sub

    Public Overrides ReadOnly Property HideRubberLine() As Boolean
        Get
            Return True
        End Get
    End Property
    Public Overrides ReadOnly Property Entity() As vdFigure
        Get
            Return line
        End Get
    End Property
    Protected Overrides Sub OnMyPositionChanged(ByVal NewPosition As gPoint)
        line.EndPoint = NewPosition
    End Sub
    Public Overrides ReadOnly Property needUpdate() As Boolean
        Get
            Return True
        End Get
    End Property

    Public Shared Sub CmdArrowLine(ByVal document As vdDocument)
        Dim EPT As gPoint, SPT As gPoint
        document.Prompt("Start Point:")
        Dim ret As Object = document.ActionUtility.getUserPoint()
        document.Prompt(Nothing)
        If ret Is Nothing Or Not (TypeOf ret Is gPoint) Then
            Return
        End If
        SPT = CType(ret, gPoint)

        Dim aFig As ActionArrowLine = New ActionArrowLine(SPT, document.ActiveLayOut)
        document.Prompt("End Point :")

        document.ActionAdd(aFig)
        Dim scode As StatusCode = aFig.WaitToFinish()
        document.Prompt(Nothing)
        If scode <> VectorDraw.Actions.StatusCode.Success Then
            Return
        End If

        EPT = CType(aFig.Value, gPoint)

        Dim line As VectorDrawArrowLine = New VectorDrawArrowLine()
        document.ActionLayout.Entities.AddItem(line)
        document.UndoHistory.PushEnable(False)
        line.InitializeProperties()
        line.setDocumentDefaults()
        line.StartPoint = SPT
        line.EndPoint = EPT
        line.Transformby(document.User2WorldMatrix)
        document.UndoHistory.PopEnable()
        document.ActionDrawFigure(line)
    End Sub
End Class

File: CustomObjects/vdPolygon/VectorDrawArrowLine.cs
﻿using System;
using System.Runtime.InteropServices;
using System.ComponentModel;
using VectorDraw.Serialize;
using VectorDraw.Professional;
using VectorDraw.Professional.vdPrimaries;
using VectorDraw.Professional.vdFigures;
using VectorDraw.Professional.vdCollections;
using VectorDraw.Professional.vdObjects;
using VectorDraw.Professional.Constants;
using VectorDraw.Geometry;
using VectorDraw.Render;
using VectorDraw.Professional.Actions;
using VectorDraw.Actions;

namespace VectorDrawCustomFigs
{
    public class VectorDrawArrowLine: vdFigure, IvdProxyFigure
    {
        
        #region private fields
        private gPoints mArrowPoints = new gPoints();
        private double mArrowSize = 1.0d;
        private gPoint mStartPoint = new gPoint();
        private gPoint mEndPoint = new gPoint();

        #endregion
        #region ctor
        public VectorDrawArrowLine()
        {
            mArrowPoints.Add(new gPoint(0.0d, 0.0d, 0.0d));
            mArrowPoints.Add(new gPoint(-1.0d, -0.1665d, 0.0d));
            mArrowPoints.Add(new gPoint(-1.0d, 0.1665d, 0.0d));
            mArrowPoints.Add(new gPoint(0.0d, 0.0d, 0.0d));
        }
        public VectorDrawArrowLine(vdDocument doc)
        {
            SetUnRegisterDocument(doc);
            setDocumentDefaults();
            mArrowPoints.Add(new gPoint(0.0d, 0.0d, 0.0d));
            mArrowPoints.Add(new gPoint(-1.0d, -0.1665d, 0.0d));
            mArrowPoints.Add(new gPoint(-1.0d, 0.1665d, 0.0d));
            mArrowPoints.Add(new gPoint(0.0d, 0.0d, 0.0d));
        }
        #endregion
        #region public properties
        //You should use the AddHistory command in your properties so any change will be writen to the history
        //The string in the AddHistory command is the name of the property exactly and should be called before the property change.
        public double arrowSize 
        {
            get 
            {
                return mArrowSize;
            }
            set 
            {
                AddHistory("arrowSize", value);
                mArrowSize = value;
            } 
        }
        public gPoint StartPoint
        {
            get
            {
                return mStartPoint;
            }
            set
            {
                AddHistory("StartPoint", value);
                mStartPoint.CopyFrom( value);
            }
        }
        public gPoint EndPoint
        {
            get
            {
                return mEndPoint;
            }
            set
            {
                AddHistory("EndPoint", value);
                mEndPoint.CopyFrom( value);
            }
        }
        #endregion
        #region private methods
        
        
        Matrix ArrowEcsMatrix(Vector ViewDir)
        {
                Matrix  mArrowMatrix = new Matrix();
                mArrowMatrix.ScaleMatrix(mArrowSize, mArrowSize, 1.0d);
                mArrowMatrix.ApplyECS2WCS(ViewDir, new Vector(StartPoint, EndPoint));
                mArrowMatrix.TranslateMatrix(EndPoint);
                return mArrowMatrix;
            
        }
        #endregion
        #region pline overrides

        //Note that not all values/objects can be saved/serialize by VDF, and you need to write code for such properties in serialize/deserialize override of this object. 
        //You can set value of custom property to any of the following type values: 
        //     String, Double, Int32, Boolean, Byte, Int16, Int64, UInt16, UInt64, UInt32, Single, ushort, System.Color, VectorDraw.Serialize.IVDHandle and IVDSerialise

        public override void Serialize(Serializer serializer)
        {
            base.Serialize(serializer);
            serializer.Serialize(mArrowSize, "arrowSize");
            serializer.Serialize(mStartPoint, "StartPoint");
            serializer.Serialize(mEndPoint, "EndPoint");
        }
        public override bool DeSerialize(DeSerializer deserializer, string fieldname, object value)
        {
            if (base.DeSerialize(deserializer, fieldname, value)) return true;
            else if (fieldname == "arrowSize") mArrowSize = (double)value;
            else if (fieldname == "StartPoint") mStartPoint = (gPoint)value;
            else if (fieldname == "EndPoint") mEndPoint = (gPoint)value;
            else return false;
            return true;
        }
        public override bool getOsnapPoints(Matrix object2viewcs, OsnapMode mode, gPoint pickPoi, gPoint LastPoi, int SegCount, OsnapPoints osnaps)
        {
            vdLine line = new vdLine();
            line.StartPoint = StartPoint;
            line.EndPoint = EndPoint;
            line.MatchProperties(this, Document);

            return line.getOsnapPoints(object2viewcs, mode, pickPoi, LastPoi, SegCount, osnaps);
        }
        public override Box BoundingBox
        {
            get
            {
                if (!mBoundBox.IsEmpty) return mBoundBox;
                mBoundBox = new Box();
                mBoundBox.AddPoint(StartPoint);
                mBoundBox.AddPoint(EndPoint);
                gPoints pts = new gPoints();
                pts.AddRange(mArrowPoints);
                ArrowEcsMatrix(new Vector(0,0,1)).Transform(pts);
                mBoundBox.AddPoints(pts);
                return mBoundBox;
                
            }
        }
        public override void Transformby(Matrix mat)
        {
            StartPoint = mat.Transform(StartPoint);
            EndPoint = mat.Transform(EndPoint);
            base.Transformby(mat);
        }
        public override vdRender.DrawStatus Draw(vdRender render)
        {
            if (ListStart(render)) return render.StatusDraw;
            vdRender.DrawStatus doDraw = base.Draw(render);
            if (doDraw == vdRender.DrawStatus.Successed)
            {
                render.DrawLine(this, StartPoint,EndPoint);
                render.PushMatrix(ArrowEcsMatrix(render.ViewDir));
                render.DrawSolidPolygon(this, mArrowPoints, vdRender.PolygonType.Simple);
                render.PopMatrix();   
            }
            AfterDraw(render);
            if (ListEnd(render)) doDraw = render.StatusDraw;
            return doDraw;
        }
        public override gPoints GetGripPoints()
        {
            gPoints ret = new gPoints();
            ret.Add(StartPoint);
            ret.Add(EndPoint);
            return ret;
        }
        public override void MoveGripPointsAt(Int32Array Indexes, double dx, double dy, double dz)
        {
            if (Indexes == null || Indexes.Count == 0 || (dx == 0.0 && dy == 0.0 && dz == 0.0)) return;
            gPoints grips = GetGripPoints();
            if (Indexes.Count == grips.Count)
            {
                Matrix mat = new Matrix();
                mat.TranslateMatrix(dx, dy, dz);
                Transformby(mat);
            }
            else
            {
                foreach (int index in Indexes)
                {
                    switch (index)
                    {
                        case 0:
                            StartPoint += new gPoint(dx, dy, dz);
                            break;
                        case 1:
                            EndPoint += new gPoint(dx, dy, dz);
                            break;
                        default:
                            break;
                    }
                }
            }
            Update();

        }
        public override void MatchProperties(VectorDraw.Professional.vdObjects.vdPrimary _from, VectorDraw.Professional.vdObjects.vdDocument thisdocument)
        {
            base.MatchProperties(_from, thisdocument);
            VectorDrawArrowLine from = _from as VectorDrawArrowLine;
            if (from == null) return;
            StartPoint = from.StartPoint;
            EndPoint = from.EndPoint;
            arrowSize = from.arrowSize;
        }
        public override vdEntities Explode()
        {
            vdEntities Entities = new vdEntities();
            Entities.SetUnRegisterDocument(Document);
            if (Document != null) Document.UndoHistory.PushEnable(false);
            vdLine line = new vdLine();
            line.StartPoint = StartPoint;
            line.EndPoint = EndPoint;
            line.MatchProperties(this,Document);
            Entities.AddItem(line);

            vdPolyline pl = new vdPolyline();
            pl.VertexList.AddRange(this.mArrowPoints);
            pl.Flag = VdConstPlineFlag.PlFlagCLOSE;
            pl.HatchProperties = new vdHatchProperties();
            pl.HatchProperties.FillMode = VdConstFill.VdFillModeSolid;
            pl.HatchProperties.FillColor.ByBlock = true;
            pl.Transformby(ArrowEcsMatrix(new Vector(0,0,1)));
            pl.MatchProperties(this,Document);
            Entities.AddItem(pl);
            if (Document != null) Document.UndoHistory.PopEnable();
            return Entities;
        }
        public override bool IsTableObjectDependOn(vdPrimary table)
        {
            //we call only the base check because there are not tables objects reference to this object properties.
            if (Deleted) return false;
            if(base.IsTableObjectDependOn(table)) return true;
            return false;
        }
        public override void GetTableDependecies(vdTableDependeciesArgs args)
        {
            //Since there no extra dependencies the base call is enough.
            base.GetTableDependecies(args);
        }
        #endregion
    }
    public class ActionArrowLine : ActionEntity
    {
        private VectorDrawArrowLine line;
        public ActionArrowLine(gPoint reference, vdLayout layout)
            : base(reference, layout)
        {
            line = new VectorDrawArrowLine();
            line.SetUnRegisterDocument(layout.Document);
            line.setDocumentDefaults();
            line.StartPoint = reference;
            line.EndPoint = reference;
        }
        public override bool HideRubberLine
        {
            get
            {
                return true;
            }
        }
        public override vdFigure Entity
        {
            get
            {
                return line;
            }
        }
        protected override void OnMyPositionChanged(gPoint newPosition)
        {
            line.EndPoint = newPosition;
        }
        public override bool needUpdate
        {
            get
            {
                return true;
            }
        }

        public static void CmdArrowLine(vdDocument document)
        {  
            gPoint EPT, SPT;
            document.Prompt("Start Point:");
            object ret = document.ActionUtility.getUserPoint();
            document.Prompt(null);
            if (ret == null || !(ret is gPoint)) return;
            SPT = ret as gPoint;

            ActionArrowLine aFig = new ActionArrowLine(SPT, document.ActiveLayOut);
            document.Prompt("End Point :");

            document.ActionAdd(aFig);
            StatusCode scode = aFig.WaitToFinish();
            document.Prompt(null);
            if (scode != VectorDraw.Actions.StatusCode.Success) return;

            EPT = aFig.Value as gPoint;

            VectorDrawArrowLine line = new VectorDrawArrowLine();
            document.ActionLayout.Entities.AddItem(line);
            document.UndoHistory.PushEnable(false);
            line.InitializeProperties();
            line.setDocumentDefaults();
            line.StartPoint = SPT;
            line.EndPoint = EPT;
            line.Transformby(document.User2WorldMatrix);
            document.UndoHistory.PopEnable();
            document.ActionDrawFigure(line);
        }
    }

}
